### PR TITLE
net/net_processing: Add thread safety related annotations for CNode and Peer

### PR DESCRIPTION
**Pull request by ajtowns**
Adds `GUARDED_BY` and `const` annotations to document how we currently ensure various members of `CNode` and `Peer` aren't subject to race conditions.

### CODE REVIEW

**Comment by ajtowns** [comment]
This breaks out the easy parts from #24474, which in turn was broken out from #21527.

File: src/net.h [hunk 1]
```diff
@@ -466,7 +466,7 @@ class CNode
      * from the wire. This cleaned string can safely be logged or displayed.
      */
     std::string cleanSubVer GUARDED_BY(m_subver_mutex){};
-    bool m_prefer_evict{false}; // This peer is preferred for eviction.
+    bool m_prefer_evict{false}; // This peer is preferred for eviction. (treated as const)
```
Comment (jnewbery) [inline src/net.h:469]: As your commit log already says, I think it'd be better just to actually make this const and included as a ctor parameter (and also made private).

**Review by jnewbery — COMMENTED** [review]
concept ACK

File: src/net_processing.cpp [hunk 2]
```diff
@@ -599,14 +599,14 @@ class PeerManagerImpl final : public PeerManager
     std::atomic<int> m_best_height{-1};
 
     /** Next time to check for stale tip */
-    std::chrono::seconds m_stale_tip_check_time{0s};
+    std::chrono::seconds m_stale_tip_check_time GUARDED_BY(cs_main){0s};
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp:602]: That it is *accessible* by other threads is why it needs to be guarded -- that's what ensures that it is only *accessed* by one thread at a time (or ever). Right now, if it were accessed by other threads who'd locked `cs_main`, it would be safe -- all this does is document that.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -599,14 +599,14 @@ class PeerManagerImpl final : public PeerManager
...
 
     /** Whether this node is running in blocks only mode */
     const bool m_ignore_incoming_txs;
 
     /** Whether we've completed initial sync yet, for determining when to turn
       * on extra block-relay-only peers. */
-    bool m_initial_sync_finished{false};
+    bool m_initial_sync_finished GUARDED_BY(cs_main){false};
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp:609]: As above, if you have `cs_main` it is safe to access this. It is not unconditionally safe for anyone to access it (unlike if it were atomic or const), hence it should be guarded. This is attempting to be a minimal, documentation only PR, so removing the variable is out of scope.

File: src/net.h [hunk 4]
```diff
@@ -413,10 +413,10 @@ class CNode
     friend struct ConnmanTestMsg;
 
 public:
-    std::unique_ptr<TransportDeserializer> m_deserializer;
-    std::unique_ptr<TransportSerializer> m_serializer;
+    const std::unique_ptr<TransportDeserializer> m_deserializer; // Used only by SocketHandler thread
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net.h:416]: The comment is so the person reading the code knows what to check for.

File: src/net.h (hunk 1 above)
Comment (ajtowns) [reply src/net.h:469]: I guess I'm okay with that in a follow up, but this patchset's been stewing for about 15 months now, so I'm a bit skeptical about increasing the scope?

**Comment by w0xlt** [comment]
Approach ACK.

File: src/net_processing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vasild) [reply src/net_processing.cpp:602]: Just because a variable is accessed under `cs_main` does not mean that it has to be protected by it. It would have to be if there was a relationship between `m_stale_tip_check_time` and some other variable that is already protected by `cs_main` and both have to be accessed together, but that is not the case.

This is the `m_stale_tip_check_time` access:

```cpp
if (now > m_stale_tip_check_time) {
    ...
    m_stale_tip_check_time = now + STALE_CHECK_INTERVAL;
}
```

If that would be called by multiple threads it would need a mutex protection. Not necessary `cs_main`. I think it is best to either leave it as is (because it is only accessed by a single thread) or introduce a new mutex for `m_stale_tip_check_time`.

Consider also a future work that tries to split `cs_main` to improve concurrency - then things have to be chopped off from `cs_main`. Needlessly adding `m_stale_tip_check_time` under it will create more work in the future.

File: src/net.h [hunk 5]
```diff
@@ -413,10 +413,10 @@ class CNode
...
+    const std::unique_ptr<const TransportSerializer> m_serializer;
 
-    NetPermissionFlags m_permissionFlags{NetPermissionFlags::None};
+    NetPermissionFlags m_permissionFlags{NetPermissionFlags::None}; // treated as const outside of fuzz tester
```
Comment (vasild) [inline src/net.h:419]: Instead of just a comment, consider actually making it `const`:

<details>
<summary>diff</summary>

```diff
diff --git i/src/net.cpp w/src/net.cpp
index e275c2964d..7bcaa6feef 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -1237,15 +1237,15 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
                              addr,
                              CalculateKeyedNetGroup(addr),
                              nonce,
                              addr_bind,
                              /*addrNameIn=*/"",
                              ConnectionType::INBOUND,
-                             inbound_onion);
+                             inbound_onion,
+                             permissionFlags);
     pnode->AddRef();
-    pnode->m_permissionFlags = permissionFlags;
     pnode->m_prefer_evict = discouraged;
     m_msgproc->InitializeNode(pnode);
 
     LogPrint(BCLog::NET, "connection from %s accepted\n", addr.ToString());
 
     {
@@ -3041,15 +3041,16 @@ ServiceFlags CConnman::GetLocalServices() const
 {
     return nLocalServices;
 }
 
 unsigned int CConnman::GetReceiveFloodSize() const { return nReceiveFloodSize; }
 
-CNode::CNode(NodeId idIn, ServiceFlags nLocalServicesIn, std::shared_ptr<Sock> sock, const CAddress& addrIn, uint64_t nKeyedNetGroupIn, uint64_t nLocalHostNonceIn, const CAddress& addrBindIn, const std::string& addrNameIn, ConnectionType conn_type_in, bool inbound_onion)
+CNode::CNode(NodeId idIn, ServiceFlags nLocalServicesIn, std::shared_ptr<Sock> sock, const CAddress& addrIn, uint64_t nKeyedNetGroupIn, uint64_t nLocalHostNonceIn, const CAddress& addrBindIn, const std::string& addrNameIn, ConnectionType conn_type_in, bool inbound_onion, NetPermissionFlags permissoin_flags)
     : m_deserializer{std::make_unique<V1TransportDeserializer>(V1TransportDeserializer(Params(), idIn, SER_NETWORK, INIT_PROTO_VERSION))},
       m_serializer{std::make_unique<V1TransportSerializer>(V1TransportSerializer())},
+      m_permissionFlags{permissoin_flags},
       m_sock{sock},
       m_connected{GetTime<std::chrono::seconds>()},
       addr(addrIn),
       addrBind(addrBindIn),
       m_addr_name{addrNameIn.empty() ? addr.ToStringIPPort() : addrNameIn},
       m_inbound_onion(inbound_onion),
diff --git i/src/net.h w/src/net.h
index 5c43c5bfb8..8701e7502a 100644
--- i/src/net.h
+++ w/src/net.h
@@ -413,13 +413,13 @@ class CNode
     friend struct ConnmanTestMsg;
 
 public:
     const std::unique_ptr<TransportDeserializer> m_deserializer; // Used only by SocketHandler thread
     const std::unique_ptr<const TransportSerializer> m_serializer;
 
-    NetPermissionFlags m_permissionFlags{NetPermissionFlags::None}; // treated as const outside of fuzz tester
+    const NetPermissionFlags m_permissionFlags;
     std::atomic<ServiceFlags> nServices{NODE_NONE};
 
     /**
      * Socket used for communication with the node.
      * May not own a Sock object (after `CloseSocketDisconnect()` or during tests).
      * `shared_ptr` (instead of `unique_ptr`) is used to avoid premature close of
@@ -582,13 +582,13 @@ public:
     std::atomic<std::chrono::microseconds> m_last_ping_time{0us};
 
     /** Lowest measured round-trip time. Used as an inbound peer eviction
      * criterium in CConnman::AttemptToEvictConnection. */
     std::atomic<std::chrono::microseconds> m_min_ping_time{std::chrono::microseconds::max()};
 
-    CNode(NodeId id, ServiceFlags nLocalServicesIn, std::shared_ptr<Sock> sock, const CAddress& addrIn, uint64_t nKeyedNetGroupIn, uint64_t nLocalHostNonceIn, const CAddress& addrBindIn, const std::string& addrNameIn, ConnectionType conn_type_in, bool inbound_onion);
+    CNode(NodeId id, ServiceFlags nLocalServicesIn, std::shared_ptr<Sock> sock, const CAddress& addrIn, uint64_t nKeyedNetGroupIn, uint64_t nLocalHostNonceIn, const CAddress& addrBindIn, const std::string& addrNameIn, ConnectionType conn_type_in, bool inbound_onion, NetPermissionFlags permission_flags = NetPermissionFlags::None);
     CNode(const CNode&) = delete;
     CNode& operator=(const CNode&) = delete;
 
     NodeId GetId() const {
         return id;
     }
diff --git i/src/test/fuzz/util.cpp w/src/test/fuzz/util.cpp
index 033c6e18d5..5bbaf903c0 100644
--- i/src/test/fuzz/util.cpp
+++ w/src/test/fuzz/util.cpp
@@ -233,13 +233,12 @@ bool FuzzedSock::IsConnected(std::string& errmsg) const
 }
 
 void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman, PeerManager& peerman, CNode& node) noexcept
 {
     const bool successfully_connected{fuzzed_data_provider.ConsumeBool()};
     const ServiceFlags remote_services = ConsumeWeakEnum(fuzzed_data_provider, ALL_SERVICE_FLAGS);
-    const NetPermissionFlags permission_flags = ConsumeWeakEnum(fuzzed_data_provider, ALL_NET_PERMISSION_FLAGS);
     const int32_t version = fuzzed_data_provider.ConsumeIntegralInRange<int32_t>(MIN_PEER_PROTO_VERSION, std::numeric_limits<int32_t>::max());
     const bool relay_txs{fuzzed_data_provider.ConsumeBool()};
 
     const CNetMsgMaker mm{0};
 
     CSerializedNetMsg msg_version{
@@ -268,13 +267,12 @@ void FillNode(FuzzedDataProvider& fuzzed_data_provider, ConnmanTestMsg& connman,
     assert(node.nVersion == version);
     assert(node.GetCommonVersion() == std::min(version, PROTOCOL_VERSION));
     assert(node.nServices == remote_services);
     CNodeStateStats statestats;
     assert(peerman.GetNodeStateStats(node.GetId(), statestats));
     assert(statestats.m_relay_txs == (relay_txs && !node.IsBlockOnlyConn()));
-    node.m_permissionFlags = permission_flags;
     if (successfully_connected) {
         CSerializedNetMsg msg_verack{mm.Make(NetMsgType::VERACK)};
         (void)connman.ReceiveMsgFrom(node, msg_verack);
         node.fPauseSend = false;
         connman.ProcessMessagesOnce(node);
         {
diff --git i/src/test/fuzz/util.h w/src/test/fuzz/util.h
index 580105e442..4d3568ff81 100644
--- i/src/test/fuzz/util.h
+++ w/src/test/fuzz/util.h
@@ -297,23 +297,25 @@ auto ConsumeNode(FuzzedDataProvider& fuzzed_data_provider, const std::optional<N
     const uint64_t keyed_net_group = fuzzed_data_provider.ConsumeIntegral<uint64_t>();
     const uint64_t local_host_nonce = fuzzed_data_provider.ConsumeIntegral<uint64_t>();
     const CAddress addr_bind = ConsumeAddress(fuzzed_data_provider);
     const std::string addr_name = fuzzed_data_provider.ConsumeRandomLengthString(64);
     const ConnectionType conn_type = fuzzed_data_provider.PickValueInArray(ALL_CONNECTION_TYPES);
     const bool inbound_onion{conn_type == ConnectionType::INBOUND ? fuzzed_data_provider.ConsumeBool() : false};
+    const NetPermissionFlags permission_flags = ConsumeWeakEnum(fuzzed_data_provider, ALL_NET_PERMISSION_FLAGS);
     if constexpr (ReturnUniquePtr) {
         return std::make_unique<CNode>(node_id,
                                        local_services,
                                        sock,
                                        address,
                                        keyed_net_group,
                                        local_host_nonce,
                                        addr_bind,
                                        addr_name,
                                        conn_type,
-                                       inbound_onion);
+                                       inbound_onion,
+                                       permission_flags);
     } else {
         return CNode{node_id,
                      local_services,
                      sock,
                      address,
                      keyed_net_group,

```
</details>

File: src/net.h (hunk 1 above)
Comment (vasild) [reply src/net.h:469]: Same as [m_permissionFlags](https://github.com/bitcoin/bitcoin/pull/25174#discussion_r880323950), better make it `const`. It is even easier than `m_permissionFlags` because there are no fuzzer changes. Or should we fuzz `m_prefer_evict` too? It would be one-line extension to the fuzzer's `ConsumeNode()` to pass one more boolean parameter from `ConsumeBool()`.

File: src/net_processing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp:602]: > Just because a variable is accessed under `cs_main` does not mean that it has to be protected by it.

That's true of everything protected by `cs_main` -- if it weren't there would be no hope of replacing cs_main by some other scheme. The issue isn't whether guarding the variable in some other way would also be safe, it's about documenting some way in which we can guarantee that the use is safe, and have the compiler maintain that guarantee for us.

File: src/net.h [hunk 6]
```diff
@@ -413,10 +413,10 @@ class CNode
     friend struct ConnmanTestMsg;
 
 public:
-    std::unique_ptr<TransportDeserializer> m_deserializer;
-    std::unique_ptr<TransportSerializer> m_serializer;
+    const std::unique_ptr<TransportDeserializer> m_deserializer; // Used only by SocketHandler thread
+    const std::unique_ptr<const TransportSerializer> m_serializer;
```
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net.h:417]: They need to be pointers because they're initialised to a subclass of the type they are.

[the remark this answers is not in the document]
Comment (vasild) [reply src/net.h:417]: The type they are can be changed: `V1TransportDeserializer m_deserializer;`, as in the diff above. The polymorphism is not needed - we don't use it via a pointer to the base class.

Before:

```cpp
const std::unique_ptr<TransportDeserializer> m_deserializer;
m_deserializer{std::make_unique<V1TransportDeserializer>(V1TransportDeserializer(...))}
```

VS

After:

```cpp
V1TransportDeserializer m_deserializer;
m_deserializer{V1TransportDeserializer(...)}
```

It looks like a pure simplification.

[the remark this answers is not in the document]
Comment (jnewbery) [reply src/net.h:417]: For context, the unique ptr to a deserializer object was added in #16202. The intention was to use a polymorphism so the deserializer could be upgraded to a p2p v2 (BIP 324) deserializer later.

I tend to agree with @vasild that we should change this to be simpler. There's been almost no progress on p2p v2 in the intervening years, and it's easy enough to change this back to a polymorphism later if required.

[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net.h:417]: The draft p2p v2 patches make use of the polymorphism, so removing it now seems like it makes things worse to me. Even if it is a good idea, it's fine for a followup, and doesn't improve thread safety so isn't necessary here.

**Review by Newman1515 — COMMENTED** [review]
Submit the approve

**Comment by hebasto** [comment]
Concept ACK.

File: src/net_processing.cpp [hunk 7]
```diff
@@ -657,7 +657,7 @@ class PeerManagerImpl final : public PeerManager
...
```
Comment (MarcoFalke) [inline src/net_processing.cpp:660]: Apart from the feedback already given, this also seems arbitrary. If you want to document symbols that happen to be under the cs_main scope, why only document this one, and not everything else, like `m_highest_fast_announce` ...

Maybe leave this hunk for a separate pull request? Otherwise I am not sure if this will move forward.

Comment (ajtowns) [reply src/net_processing.cpp:660]: `m_highest_fast_announce` is a member of PeerManagerImpl, not CNode or Peer, so is out of scope for this PR...

File: src/net.h [hunk 8]
```diff
@@ -613,7 +613,7 @@ class CNode
      * closed.
      * Otherwise this unique_ptr is empty.
      */
-    std::unique_ptr<i2p::sam::Session> m_i2p_sam_session GUARDED_BY(m_sock_mutex);
+    std::unique_ptr<i2p::sam::Session> m_i2p_sam_session GUARDED_BY(m_sock_mutex) PT_GUARDED_BY(m_sock_mutex);
```
Comment (jonatack) [inline src/net.h:616]: Question, are both guards useful here? (All but one of the other instances of `PT_GUARDED_BY` don't cumulate both annotations.)  

```diff
-    std::unique_ptr<i2p::sam::Session> m_i2p_sam_session GUARDED_BY(m_sock_mutex) PT_GUARDED_BY(m_sock_mutex);
+    std::unique_ptr<i2p::sam::Session> m_i2p_sam_session PT_GUARDED_BY(m_sock_mutex);
 };
```

File: src/net.h (hunk 1 above)
Comment (jonatack) [reply src/net.h:396]: Agree to make these private data members (consider to prefer [non-const](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c12-dont-make-data-members-const-or-references)), if they do not need to be public or protected for testing/fuzzing.

**Review by jonatack — COMMENTED** [review]
ACK first commit "net/net_processing: add missing thread safety annotations" modulo the `GUARDED_BY(cs_main)` annotations

File: src/net.h (hunk 8 above)
Comment (ajtowns) [reply src/net.h:616]: No, the original annotation was correct; for some reason I thought PT_GUARDED_BY added some checks GUARDED_BY doesn't have, but that doesn't seem to be the case.

Changing from GUARDED_BY to PT_GUARDED_BY alone would allow you to do `m_i2p_sam_session = nullptr` without holding the lock, or `if (m_i2p_sam_session != nullptr) { LOCK(m_sock_mutex); m_i2p_sam_session->foo(); }` which would be racy.

File: src/net.h (hunk 1 above)
Comment (ajtowns) [reply src/net.h:396]: Making data members private does not make them thread safe. This member is thread safe in the current code, despite the lack of a guarding mutex, because its value is fixed immediately after creation, before another thread can access it, and it does not change until it's destructed.

Comment (jonatack) [reply src/net.h:396]: Yes, the point is to encapsulate if the data member participates in the object’s invariant, in which case it should probably either be const (only if thread-safe) or private.

File: src/net_processing.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajtowns) [reply src/net_processing.cpp:651]: When I look at both `m_stale_tip_check_time` and `m_initial_sync_finished` and want to convince myself that there's no chance of one thread racing to access them while another is in the middle of modifying them, the way I convince myself of that is "oh, they're both used from precisely one function, and that function just locked cs_main".

It's certainly possible to look further, and say that `CheckForStaleTipAndEvictPeers` is only called via the scheduler thread; but that's more work, and it's not something that I can easily get the compiler to check for me. I *could* tell the compiler to check it for me with more work; but that involves adding thread-specific mutexes which reviewers here have objected to previously https://github.com/bitcoin/bitcoin/pull/21527#discussion_r631895780 and the whole point of breaking this PR out was to try to make some incremental progress in the meantime.

File: src/net.h (hunk 1 above)
Comment (ajtowns) [reply src/net.h:396]: The reason it's thread-safe is that it's (effectively) const; the reason it's (effectively) const is that its value is known at the time the connection is setup. This PR is only about documenting why data members are thread safe, not changing logic or encapsulation...

I don't really get the `private` suggestion -- it's a `CNode` data member, but it's not accessed by any `CNode` member functions, it's only used by `CConnman::AttemptToEvictConnection`. While `CConnman` is a friend class so could get away with that, it seems like a weird thing to do.

https://github.com/ajtowns/bitcoin/commits/202208-cnodeoptions has a commit with these actually made `const` for whatever that's worth. Again, it does more than just document the current state of the code, so I'm not adding it to this PR.

**Comment by MarcoFalke** [comment]
review ACK 9816dc96b77fd9780bb97891cd45a1b9798db8f5 📍

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 9816dc96b77fd9780bb97891cd45a1b9798db8f5 📍
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiGPQwAk8l83FaAio2oGaavZm/DW+AtDfrRktDHvmfZC78JmukdBl+ZFGYStrY+
b7WdSH1n98Xf1Czr1oHrMc1uEHxj/ajAw0lR3ELurht0jRv6o12YMiltI6E21olh
YFhrHVDPwPOgfcS8W1quNAv21MFX1tQTP7vFyyHiZbkP51dHlLKOJRah6SL9Ag9l
CyFXjrDtIE6bgOF06+/T/3KFPDxbxy8vrhsMTGeg33oj6FcCvP+xhkLrwyrmdHnT
mmcHCjfmoWRHcY5alEKOUQAFw+olpEzlLKq8mwsZoCt4prbk8tfihKu3If7LxscP
uF4dDGD3nWiOL76mVgF27MMurdIY7yCI5IHNPOzBxQX8dfjuKY0OnPXLWFJjKOXf
i2MOBQyxJ4fAkZ+KVyBo5qEh2PuwTA+oMFbIyNhaj/IDpOv9W9dHb7jqEkgVEynb
BtmQztp7SkutBiv1rfarRRSWKRkrussJzMY43U7lFe+ClZFXMEw3NJh/AqtThtdV
mG/YSC0D
=34Gj
-----END PGP SIGNATURE-----
```


</details>

**Comment by ajtowns** [comment]
Here's one way of documenting "This member is only accessed by the scheduler thread" with (I believe) no runtime impact:

```c++
class LOCKABLE CodeInspectionRequired { };
class SCOPED_LOCKABLE CodeInspectionSaysSafeToUse
{
public:
    CodeInspectionSaysSafeToUse(const CodeInspectionRequired& mut) EXCLUSIVE_LOCK_FUNCTION(mut) { }
    ~CodeInspectionSaysSafeToUse() UNLOCK_FUNCTION() { }
};
#define SAFE_TO_USE(mut) CodeInspectionSaysSafeToUse UNIQUE_NAME(criticalblock)(mut)
```

You'd then use it to annotate variables where you have to manually inspect the logic used to access them:

```c++
    static constexpr CodeInspectionRequired FROM_SCHEDULER_THREAD;
    std::chrono::seconds m_stale_tip_check_time GUARDED_BY(FROM_SCHEDULER_THREAD){0s};
```

but could pass that logic up through functions:

```c++
    void CheckForStaleTipAndEvictPeers() EXCLUSIVE_LOCKS_REQUIRED(FROM_SCHEDULER_THREAD) override;
```

to the caller:

```c++
    scheduler.scheduleEvery([this] { SAFE_TO_USE(FROM_SCHEDULER_THREAD); this->CheckForStaleTipAndEvictPeers(); }, std::chrono::seconds{EXTRA_PEER_CHECK_INTERVAL});
```

In this particular case, you'd need to put the `FROM_SCHEDULER_THREAD` declaration in the `PeerManager` virtual class where `CheckForStaleTipAndEvictPeers` is first declared there, and also add `SAFE_TO_USE` annotations in the unit tests. Without the `SAFE_TO_USE` annotation, clang errors with:

```
net_processing.cpp:1695:81: error: calling function 'CheckForStaleTipAndEvictPeers' requires holding mutex 'FROM_SCHEDULER_THREAD' exclusively [-Werror,-Wthread-safety-analysis]
    scheduler.scheduleEvery([this] { this->CheckForStaleTipAndEvictPeers(); }, std::chrono::seconds{EXTRA_PEER_CHECK_INTERVAL});
```

Is something like the above a good idea? It seems better than what we have now, where we don't document why a variable is safe at all, and could introduce a race by having one PR introduce a new access from another thread with `cs_main` held, while another introduces a new access from the scheduler thread without `cs_main` held, with a bug introduced because neither reviewers nor the compiler hasn't been told what to expect.

But compared to just saying `GUARDED_BY(cs_main)` when it already is guarded by cs_main, the above seems needlessly complicated and like it puts too much work on developers rather than tooling.

**Comment by jonatack** [comment]
utACK 9816dc96b77fd9780bb97891cd45a1b9798db8f5

**Review by hebasto — APPROVED** [review]
ACK 9816dc96b77fd9780bb97891cd45a1b9798db8f5, I have reviewed the code and it looks OK. In particular, I verified the usage of variables which got `GUARDED_BY` annotations.

Still curious about PR author's opinion about https://github.com/bitcoin/bitcoin/pull/25174#discussion_r880323950.

**Comment by MarcoFalke** [comment]
> Still curious about PR author's opinion about https://github.com/bitcoin/bitcoin/pull/25174#discussion_r880323950.

See https://github.com/bitcoin/bitcoin/pull/25174#discussion_r957281170

File: src/net.h (hunk 1 above)
Comment (ajtowns) [reply src/net.h:396]: Made const in #25962
